### DISCUSSION TITLE
Add healthcheck to docker image for better docker management

### DISCUSSION THREAD

**Discussion by benjibenoit** [directus/directus#12268] [Feature Requests]
Hi,

is it possible to think about adding an endpoint healthcheck on the docker image? Using portainer I would like to have a healthcheck to manage my containers better ?

Regards,
Benjamin

**Comment by azrikahar** [comment]
You should be able to leverage the existing `/server/health` endpoint for this 👍 It'll also show more detail if you're authenticated, more info on it in the docs: https://docs.directus.io/reference/system/server/#health

**Reply by benjibenoit** [reply]
I know that Directus has a healthcheck endpoint. I was talking about a docker endpoint. So that Docker can know what the status is. I take for example traefik which can check if a container is up. You have to say, on the Dockerfile, how to manage the healthcheck with the HEALTHCHECK entry

**Reply by azrikahar** [reply]
Thanks for clarifying! I personally got used to configuring it in docker compose rather than the Dockerfile itself, hence that's what I was thinking about when mentioning the existing endpoint. Could be a good enhancement indeed. 

In the meantime, if you would like to use the Dockerfile route, perhaps you can extend the existing Directus docker image with something like:

```Dockerfile
HEALTHCHECK --interval=15s --timeout=15s --start-period=30s \  
    CMD node healthcheckscript.js
```

where `healthcheckscript.js` contains a minimal code that fetches that endpoint. Of course there's always the simpler alternative of just installing and use `curl` instead.

**Reply by benjibenoit** [reply]
Oh, I didn't know you could do that directly in docker-compose. So I'll add that right away!
```dockerfile
    healthcheck:
      test: [ "CMD", "curl", "--fail", "<ip-address>", "||", "exit", "1" ]
      interval: 15s
      timeout: 15s
```

**Reply by benjibenoit** [reply]
um, I see that directus doesn't have curl, so I'm using wget:
```dockerfile
    healthcheck:
      test: [ "CMD", "wget", "--no-verbose", "--tries=1", "--spider", "<ip-address>", "||", "exit", "1" ]
      interval: 15s
      timeout: 15s
```

**Reply by benjibenoit** [reply]
Well, when I put this on, my container doesn't want to launch anymore and I have this error:
```shell
npm ERR! path /directus
npm ERR! command failed
npm ERR! signal SIGTERM
npm ERR! command sh -c directus "start"
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_logs/2022-03-21T22_26_26_585Z-debug-0.log

```

When I look in the log file I see this, what do you say? 
```shell
29 verbose stack Error: command failed
29 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/index.js:64:27)
29 verbose stack     at ChildProcess.emit (node:events:520:28)
29 verbose stack     at maybeClose (node:internal/child_process:1092:16)
29 verbose stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)
30 verbose cwd /directus
31 verbose Linux 5.4.0-96-generic
32 verbose argv "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/bin/npm-cli.js" "exec" "--" "directus" "start"
33 verbose node v16.14.0
34 verbose npm  v8.3.1
35 error path /directus
36 error command failed
37 error signal SIGTERM
38 error command sh -c directus "start"
39 verbose exit 1
40 timing npm Completed in 42177ms
```

**Reply by azrikahar** [reply]
> um, I see that directus doesn't have curl, so I'm using wget

Technically it's `node:alpine` that doesn't have curl, which is the image Directus is based off, but that is indeed the reason why I didn't suggest curl initially and mentioned "**installing** and use curl", emphasis on "installing" 😄 but `wget` is definitely the right call here!

> Well, when I put this on, my container doesn't want to launch anymore and I have this error

I believe that is actually unrelated to the healthcheck setup, as I did several attempts where although my healthcheck config was not correct, the container still runs as usual. You should check your other configurations like volume mounts, and make sure not to override the default command etc.

With that said, I did test the healthcheck config and tweak it abit so it works. I'd recommend trying the following minimal compose file to see does it work for you or not:

(feel free to change `8099` to `8055`. I changed it just to avoid port conflict on my machine)

```yaml
services:
    directus:
        image: directus/directus:9.7.0
        container_name: directus
        restart: unless-stopped
        ports:
            - 8099:8055
        environment:
            - KEY=thisisasuperkey
            - SECRET=thisisasupersecret
            - ADMIN_EMAIL=admin@example.com
            - ADMIN_PASSWORD=password
        healthcheck:
            test: wget --no-verbose --tries=1 --spider http://localhost:8055/server/health || exit 1
            start_period: 15s
            interval: 15s
            timeout: 15s
            retries: 3
```

Basically the changes I did are:

- Ended up having to change `test:` out of the bracket syntax to make it report healthy properly. Honestly I'm not sure what's the exact reasoning since I also rather seldom uses it, but maybe bracket syntax requires some changes that I don't know of.

- added `start_period` and `retries`, probably not necessary, but good to have nonetheless (since migrations may make it not responsive for a hot second in the beginning)

- Not sure if you're using this, but do make sure it's `http://localhost:8055/server/health` for your `<ip-address>` argument for wget.

Sample output of it working:

![WindowsTerminal_FMNIcLQuni](https://user-images.githubusercontent.com/42867097/159424338-2f3bd302-4a9f-46bf-b00d-c2ae91bea3db.png)

![WindowsTerminal_SFibNhGx0G](https://user-images.githubusercontent.com/42867097/159424359-1c696cc9-6d7f-4b64-9057-78a5cf60ce7e.png)

**Reply by benjibenoit** [reply]
It seems to work well that way! Now I see that I have a warning that my s3 is taking too long to respond. I guess that means that the Healthcheck is working :D

**Reply by nodeworks** [reply]
I'm also getting that SIGTERM error as @shankiflang did. @shankiflang did you ever get the problem coming back after you made the change above?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by ravyshankar** [comment]
I have the same problem. Curl is not installed and I dont know how to healtcheck the ECS container running the directus image. Wish directus provides better documentation with the necessary pre-requisites.

**Reply by rijkvanzanten** [reply]
There's a million different ways on how you can host containers / apps on various platforms with different requirements and gotchas. There's no way for me to document them all, or even keep a handful up to date with the many changes in the hosting landscape.
